### PR TITLE
Added test_chunk and fixed

### PR DESCRIPTION
**Pull request by ZenginU**
Added test_chunk and changed the implementation because previous one fails the tests. 

Test cases are bit mess right now as I tried to figure out how to test for mod != 0 and dimension tests.

I have changed the implementation (checks were failing).

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/tensor.py                         4               4
------------------------------------------------------------
lines added in the tinygrad folder: 0
```

**Comment by geohot** [comment]
"previous one fails the tests"

What exactly fails and why?

https://thoughtbot.com/blog/chestertons-fence

**Comment by ZenginU** [comment]
@geohot

Firstly, unfortunately there was no previous test_chunk. 
Secondly, I believe previous code is aimed to split the kernel perfectly.

`If the tensor size along the given dimension dim is not divisible by chunks, all returned chunks will be the same size, except the last one`
Previous code lacks this condition so when you try to split 14 into 3 chunk code fails.

I hope this explanation is sufficient

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -341,10 +341,11 @@ def repeat(self, repeats):
 
   # TODO: make this nicer with syntactic sugar in slice
   def chunk(self, num, dim):
-    slice_params = [[(0, s) for s in self.shape] for _ in range(num)]
-    for i,k in enumerate(range(0, self.shape[dim], self.shape[dim]//num)):
-      slice_params[i][dim] = (k, min(self.shape[dim], k+self.shape[dim]//num))
-    return [self.slice(p) for p in slice_params]
+    dim = (dim + self.ndim) if dim < 0 else dim
+    x = self.permute(*(i for i in range(self.ndim) if i != dim), dim)
```
Comment (geohot) [inline tinygrad/tensor.py:345]: Why is this permuting?

Comment (ZenginU) [reply tinygrad/tensor.py:345]: In order not to deal with slice_params, I moved the dimension that will be chunked to last dim to use the "..." in the for loop. May be there is a more optimized approach, if you want I can try to find it

Comment (geohot) [reply tinygrad/tensor.py:345]: Yea, do it the right way

**Comment by ZenginU** [comment]
@geohot 

So made the code shorter and not used Tensor.slice as it is discouraged,

But if you say that this code is unreadable I have find the bug in the old code as well so I can just change it preserving the readability

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -341,10 +341,9 @@ def repeat(self, repeats):
 
   # TODO: make this nicer with syntactic sugar in slice
   def chunk(self, num, dim):
-    slice_params = [[(0, s) for s in self.shape] for _ in range(num)]
-    for i,k in enumerate(range(0, self.shape[dim], self.shape[dim]//num)):
```
Comment (ZenginU) [inline tinygrad/tensor.py:345]: changing
```python
self.shape[dim]//num -> ceil(self.shape[dim]/num)
```  
in all occurences fixes the bug in prev code, it was late for me to realize this unfortunately.

Comment (wozeparrot) [reply tinygrad/tensor.py:345]: I think the old chunk code is more readable, maybe try the fix on the old code and see if it passes the test?

Comment (ZenginU) [reply tinygrad/tensor.py:345]: Hello @wozeparrot,

Most of it looking the same only using slice(). But if still unreadable than I would directly update the old chunk code.

I am insisting on not using tensor.slice but instead using builtin slice(). Sometimes tensor.slice() can add unexpected zero padding.

Comment (wozeparrot) [reply tinygrad/tensor.py:345]: It's much more readable now. :)

**Comment by wozeparrot** [comment]
Can you add a test for negative dims?

**Comment by ZenginU** [comment]
Added new tests but testing seems odd to me but cant think any other way as .chunk returns list. Any ideas on this?

**Comment by wozeparrot** [comment]
You can try to add a special case in `helper_test_op` but i don't think it will be much cleaner than this, cuz this is already pretty clean.

**Comment by ZenginU** [comment]
@wozeparrot 
Find more bugs related with old code, first line became bit ugly (inside the range()) but I hope gets the job done and ready for merge

File: tinygrad/tensor.py [hunk 3]
```diff
@@ -341,10 +341,10 @@ def repeat(self, repeats):
...
-      slice_params[i][dim] = (k, min(self.shape[dim], k+self.shape[dim]//num))
-    return [self.slice(p) for p in slice_params]
+    slice_params = [[slice(None) for s in self.shape] for _ in range(ceil(self.shape[dim]/ceil(self.shape[dim]/num)))]
```
Comment (wozeparrot) [inline tinygrad/tensor.py:344]: Do you need to initialize with `slice(None)` here? does initializing with just `None` work?

Comment (ZenginU) [reply tinygrad/tensor.py:344]: Replacing slice(None) with None breaks, my idea with the slice(None) is ":" but i think None acts different as it throws `AssertionError: zeros not allowed in shape (1, 0, 13)`

Btw, I have plans to replace the slice with .pad in .pad2d as well should I do it? Maybe I can combine them in this PR as well as it will be small

Comment (wozeparrot) [reply tinygrad/tensor.py:344]: Should keep that in a separate PR. I'll just merge this one now.
